File: backend/parser/parser.py
"""Parser that converts a token stream into a ProcessModel."""

from models.fpb_model import (
    Flow,
    FlowType,
    Identification,
    ProcessOperator,
    State,
    StatePlacement,
    StateType,
    TechnicalResource,
    Usage,
)
from models.fpb_model import SystemLimit
from models.process_model import ProcessModel
from parser.lexer import Lexer, Token
from parser.syntax import ELEMENT_KEYWORDS, TokenType


class ParseError(Exception):
    """Error raised during parsing."""

    def __init__(self, message: str, line: int = 0, column: int = 0) -> None:
        self.line = line
        self.column = column
        super().__init__(f"Line {line}, Col {column}: {message}")


# Maps element keywords to StateType values
_STATE_KEYWORD_MAP = {
    "product": StateType.PRODUCT,
    "energy": StateType.ENERGY,
    "information": StateType.INFORMATION,
}

# Maps connection token types to FlowType values
_FLOW_TOKEN_MAP = {
    TokenType.FLOW: FlowType.FLOW,
    TokenType.ALTERNATIVE_FLOW: FlowType.ALTERNATIVE_FLOW,
    TokenType.PARALLEL_FLOW: FlowType.PARALLEL_FLOW,
}


class FpbParser:
    """Parser that converts FPB text into a ProcessModel."""

    def __init__(self, source: str) -> None:
        self.source = source
        self.tokens: list[Token] = []
        self.pos = 0
        self.model = ProcessModel()
        self._element_ids: dict[str, str] = {}  # id -> element type keyword
        self._flow_counter = 0
        self._usage_counter = 0
        self._current_system_id: str | None = None
        self._system_counter = 0

    def parse(self) -> ProcessModel:
        """Parse the source text and return a ProcessModel."""
        self.tokens = Lexer(self.source).tokenize()
        self.pos = 0

        self._skip_trivial()
        self._expect(TokenType.START_FPB)
        self._skip_trivial()

        while not self._check(TokenType.END_FPB) and not self._check(TokenType.EOF):
            self._parse_statement()
            self._skip_trivial()

        if self._check(TokenType.END_FPB):
            self._advance()
        else:
            self.model.errors.append("Missing @endfpb delimiter")

        return self.model

    # -- Statement dispatch --

    def _parse_statement(self) -> None:
        """Parse a single statement."""
        token = self._current()

        if token.type == TokenType.COMMENT:
            self._advance()
            return

        if token.type == TokenType.KEYWORD:
            if token.value == "title":
                self._parse_title()
            elif token.value == "system":
                self._parse_system_block()
            elif token.value in ELEMENT_KEYWORDS:
                self._parse_element_decl()
            else:
                self.model.errors.append(
                    f"Line {token.line}: Unknown keyword '{token.value}'"
                )
                self._advance()
            return

        if token.type == TokenType.IDENTIFIER:
            self._parse_connection()
            return

        # Skip unexpected tokens
        self._advance()

    # -- Title --

    def _parse_title(self) -> None:
        """Parse: title STRING"""
        if self._current_system_id is not None:
            self.model.errors.append(
                f"Line {self._current().line}: title cannot be used inside a system block"
            )
            self._advance()
            return
        self._advance()  # consume 'title'
        if self._check(TokenType.STRING):
            self.model.title = self._current().value
            self._advance()
        else:
            self.model.errors.append(
                f"Line {self._current().line}: Expected string after 'title'"
            )

    # -- System blocks --

    def _parse_system_block(self) -> None:
        """Parse: system STRING { Statement* }"""
        system_token = self._current()
        self._advance()  # consume 'system'

        if not self._check(TokenType.STRING):
            self.model.errors.append(
                f"Line {system_token.line}: Expected string after 'system'"
            )
            return

        system_name = self._current().value
        self._advance()

        if not self._check(TokenType.LBRACE):
            self.model.errors.append(
                f"Line {system_token.line}: Expected '{{' after system name"
            )
            return
        self._advance()  # consume '{'

        self._system_counter += 1
        system_id = f"system_{self._system_counter}"

        ident = Identification(unique_ident=system_id, long_name=system_name)
        self.model.system_limits.append(
            SystemLimit(
                id=system_id,
                identification=ident,
                label=system_name,
                line_number=system_token.line,
            )
        )

        self._current_system_id = system_id
        self._skip_trivial()

        while not self._check(TokenType.RBRACE) and not self._check(TokenType.EOF) and not self._check(TokenType.END_FPB):
            self._parse_statement()
            self._skip_trivial()

        if self._check(TokenType.RBRACE):
            self._advance()
        else:
            self.model.errors.append(
                f"Line {system_token.line}: Missing '}}' for system '{system_name}'"
            )

        self._current_system_id = None

    # -- Element declarations --

    def _parse_element_decl(self) -> None:
        """Parse: element_type IDENTIFIER STRING? ANNOTATION?"""
        keyword_token = self._current()
        keyword = keyword_token.value
        self._advance()  # consume keyword

        if not self._check(TokenType.IDENTIFIER):
            self.model.errors.append(
                f"Line {keyword_token.line}: Expected identifier after '{keyword}'"
            )
            return

        elem_id = self._current().value
        self._advance()

        # Check for duplicate ID
        if elem_id in self._element_ids:
            self.model.errors.append(
                f"Line {keyword_token.line}: Duplicate element ID '{elem_id}'"
            )
            return

        # Optional label
        label = elem_id
        if self._check(TokenType.STRING):
            label = self._current().value
            self._advance()

        # Optional placement annotation
        placement = None
        if self._check(TokenType.ANNOTATION):
            annotation_value = self._current().value
            self._advance()
            if keyword in _STATE_KEYWORD_MAP:
                _ANNOTATION_MAP = {
                    "@boundary": StatePlacement.BOUNDARY,
                    "@boundary-top": StatePlacement.BOUNDARY_TOP,
                    "@boundary-bottom": StatePlacement.BOUNDARY_BOTTOM,
                    "@boundary-left": StatePlacement.BOUNDARY_LEFT,
                    "@boundary-right": StatePlacement.BOUNDARY_RIGHT,
                    "@internal": StatePlacement.INTERNAL,
                }
                placement = _ANNOTATION_MAP.get(annotation_value)
            else:
                self.model.warnings.append(
                    f"Line {keyword_token.line}: Placement annotation "
                    f"'{annotation_value}' ignored on '{keyword}' "
                    f"(only valid on state elements)"
                )

        ident = Identification(unique_ident=elem_id, long_name=label)
        self._element_ids[elem_id] = keyword

        if keyword in _STATE_KEYWORD_MAP:
            self.model.states.append(
                State(
                    id=elem_id,
                    state_type=_STATE_KEYWORD_MAP[keyword],
                    identification=ident,
                    label=label,
                    placement=placement,
                    line_number=keyword_token.line,
                    system_id=self._current_system_id,
                )
            )
        elif keyword == "process_operator":
            self.model.process_operators.append(
                ProcessOperator(
                    id=elem_id,
                    identification=ident,
                    label=label,
                    line_number=keyword_token.line,
                    system_id=self._current_system_id,
                )
            )
        elif keyword == "technical_resource":
            self.model.technical_resources.append(
                TechnicalResource(
                    id=elem_id,
                    identification=ident,
                    label=label,
                    line_number=keyword_token.line,
                    system_id=self._current_system_id,
                )
            )

    # -- Connections --

    def _parse_connection(self) -> None:
        """Parse: IDENTIFIER connector IDENTIFIER"""
        source_token = self._current()
        source_id = source_token.value
        self._advance()

        token = self._current()
        if token.type == TokenType.USAGE:
            self._advance()
            if not self._check(TokenType.IDENTIFIER):
                self.model.errors.append(
                    f"Line {source_token.line}: Expected identifier after connection operator"
                )
                return
            target_id = self._current().value
            self._advance()

            if source_id not in self._element_ids:
                self.model.errors.append(
                    f"Line {source_token.line}: Element '{source_id}' is not defined"
                )
                return
            if target_id not in self._element_ids:
                self.model.errors.append(
                    f"Line {source_token.line}: Element '{target_id}' is not defined"
                )
                return

            self._usage_counter += 1
            self.model.usages.append(
                Usage(
                    id=f"usage_{self._usage_counter}",
                    process_operator_ref=source_id,
                    technical_resource_ref=target_id,
                    line_number=source_token.line,
                    system_id=self._current_system_id,
                )
            )

        elif token.type in _FLOW_TOKEN_MAP:
            flow_type = _FLOW_TOKEN_MAP[token.type]
            self._advance()

            if not self._check(TokenType.IDENTIFIER):
                self.model.errors.append(
                    f"Line {source_token.line}: Expected identifier after connection operator"
                )
                return
            target_id = self._current().value
            self._advance()

            if source_id not in self._element_ids:
                self.model.errors.append(
                    f"Line {source_token.line}: Element '{source_id}' is not defined"
                )
                return
            if target_id not in self._element_ids:
                self.model.errors.append(
                    f"Line {source_token.line}: Element '{target_id}' is not defined"
                )
                return

            self._flow_counter += 1
            self.model.flows.append(
                Flow(
                    id=f"flow_{self._flow_counter}",
                    source_ref=source_id,
                    target_ref=target_id,
                    flow_type=flow_type,
                    line_number=source_token.line,
                    system_id=self._current_system_id,
                )
            )

        else:
            self.model.errors.append(
                f"Line {source_token.line}: Expected connection operator after '{source_id}'"
            )

    # -- Token helpers --

    def _current(self) -> Token:
        """Return the current token."""
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return Token(TokenType.EOF, "", 0, 0)

    def _check(self, token_type: TokenType) -> bool:
        """Check if current token matches the given type."""
        return self._current().type == token_type

    def _advance(self) -> Token:
        """Advance to next token and return the consumed one."""
        token = self._current()
        if self.pos < len(self.tokens):
            self.pos += 1
        return token

    def _expect(self, token_type: TokenType) -> Token:
        """Consume a token of the expected type or record an error."""
        if self._check(token_type):
            return self._advance()
        token = self._current()
        self.model.errors.append(
            f"Line {token.line}: Expected {token_type.value}, got {token.type.value}"
        )
        return token

    def _skip_trivial(self) -> None:
        """Skip newlines and comments."""
        while self._current().type in (TokenType.NEWLINE, TokenType.COMMENT):
            self._advance()

File: packages/core/src/parser/parser.ts
/** Parser that converts a token stream into a ProcessModel. */

import {
    Flow,
    FlowType,
    Identification,
    ProcessOperator,
    State,
    StatePlacement,
    SystemLimit,
    TechnicalResource,
    Usage,
} from '../models/fpdModel';
import { STATE_TYPE_MAP as STATE_KEYWORD_MAP } from '../models/constants';
import { ProcessModel, createProcessModel } from '../models/processModel';
import { Lexer, Token } from './lexer';
import { ELEMENT_KEYWORDS, TokenType } from './syntax';

export class ParseError extends Error {
    line: number;
    column: number;

    constructor(message: string, line: number = 0, column: number = 0) {
        super(`Line ${line}, Col ${column}: ${message}`);
        this.line = line;
        this.column = column;
    }
}

/** Maps connection token types to FlowType values */
const FLOW_TOKEN_MAP: ReadonlyMap<TokenType, FlowType> = new Map([
    [TokenType.FLOW, 'flow'],
    [TokenType.ALTERNATIVE_FLOW, 'alternativeFlow'],
    [TokenType.PARALLEL_FLOW, 'parallelFlow'],
]);

/** Maps annotation strings to StatePlacement values */
const ANNOTATION_MAP: Record<string, StatePlacement> = {
    '@boundary': 'boundary',
    '@boundary-top': 'boundary-top',
    '@boundary-bottom': 'boundary-bottom',
    '@boundary-left': 'boundary-left',
    '@boundary-right': 'boundary-right',
    '@internal': 'internal',
};

export class FpdParser {
    private source: string;
    private tokens: Token[] = [];
    private pos = 0;
    private model: ProcessModel = createProcessModel();
    private elementIds: Map<string, string> = new Map(); // id -> element type keyword
    private flowCounter = 0;
    private usageCounter = 0;
    private currentSystemId: string | null = null;
    private systemCounter = 0;

    constructor(source: string) {
        this.source = source;
    }

    parse(): ProcessModel {
        this.tokens = new Lexer(this.source).tokenize();
        this.pos = 0;

        this.skipTrivial();
        this.expect(TokenType.START_FPD);
        this.skipTrivial();

        while (!this.check(TokenType.END_FPD) && !this.check(TokenType.EOF)) {
            this.parseStatement();
            this.skipTrivial();
        }

        if (this.check(TokenType.END_FPD)) {
            this.advance();
        } else {
            this.model.errors.push('Missing @endfpd delimiter');
        }

        return this.model;
    }

    // -- Statement dispatch --

    private parseStatement(): void {
        const token = this.current();

        if (token.type === TokenType.COMMENT) {
            this.advance();
            return;
        }

        if (token.type === TokenType.KEYWORD) {
            if (token.value === 'title') {
                this.parseTitle();
            } else if (token.value === 'system') {
                this.parseSystemBlock();
            } else if (ELEMENT_KEYWORDS.has(token.value)) {
                this.parseElementDecl();
            } else {
                this.model.errors.push(`Line ${token.line}: Unknown keyword '${token.value}'`);
                this.advance();
            }
            return;
        }

        if (token.type === TokenType.IDENTIFIER) {
            this.parseConnection();
            return;
        }

        // Skip unexpected tokens
        this.advance();
    }

    // -- Title --

    private parseTitle(): void {
        if (this.currentSystemId !== null) {
            this.model.errors.push(
                `Line ${this.current().line}: title cannot be used inside a system block`,
            );
            this.advance();
            return;
        }
        this.advance(); // consume 'title'
        if (this.check(TokenType.STRING)) {
            this.model.title = this.current().value;
            this.advance();
        } else {
            this.model.errors.push(`Line ${this.current().line}: Expected string after 'title'`);
        }
    }

    // -- System blocks --

    private parseSystemBlock(): void {
        const systemToken = this.current();
        this.advance(); // consume 'system'

        if (!this.check(TokenType.STRING)) {
            this.model.errors.push(`Line ${systemToken.line}: Expected string after 'system'`);
            return;
        }

        const systemName = this.current().value;
        this.advance();

        if (!this.check(TokenType.LBRACE)) {
            this.model.errors.push(`Line ${systemToken.line}: Expected '{' after system name`);
            return;
        }
        this.advance(); // consume '{'

        this.systemCounter++;
        const systemId = `system_${this.systemCounter}`;

        const ident: Identification = { uniqueIdent: systemId, longName: systemName };
        this.model.systemLimits.push({
            id: systemId,
            identification: ident,
            label: systemName,
            lineNumber: systemToken.line,
        } as SystemLimit);

        this.currentSystemId = systemId;
        this.skipTrivial();

        while (
            !this.check(TokenType.RBRACE) &&
            !this.check(TokenType.EOF) &&
            !this.check(TokenType.END_FPD)
        ) {
            this.parseStatement();
            this.skipTrivial();
        }

        if (this.check(TokenType.RBRACE)) {
            this.advance();
        } else {
            this.model.errors.push(
                `Line ${systemToken.line}: Missing '}' for system '${systemName}'`,
            );
        }

        this.currentSystemId = null;
    }

    // -- Element declarations --

    private parseElementDecl(): void {
        const keywordToken = this.current();
        const keyword = keywordToken.value;
        this.advance(); // consume keyword

        if (!this.check(TokenType.IDENTIFIER)) {
            this.model.errors.push(
                `Line ${keywordToken.line}: Expected identifier after '${keyword}'`,
            );
            return;
        }

        const elemId = this.current().value;
        this.advance();

        // Check for duplicate ID
        if (this.elementIds.has(elemId)) {
            this.model.errors.push(`Line ${keywordToken.line}: Duplicate element ID '${elemId}'`);
            return;
        }

        // Optional label
        let label = elemId;
        if (this.check(TokenType.STRING)) {
            label = this.current().value;
            this.advance();
        }

        // Optional placement annotation
        let placement: StatePlacement | undefined;
        if (this.check(TokenType.ANNOTATION)) {
            const annotationValue = this.current().value;
            this.advance();
            if (keyword in STATE_KEYWORD_MAP) {
                placement = ANNOTATION_MAP[annotationValue];
            } else {
                this.model.warnings.push(
                    `Line ${keywordToken.line}: Placement annotation ` +
                        `'${annotationValue}' ignored on '${keyword}' ` +
                        `(only valid on state elements)`,
                );
            }
        }

        const ident: Identification = { uniqueIdent: elemId, longName: label };
        this.elementIds.set(elemId, keyword);

        if (keyword in STATE_KEYWORD_MAP) {
            this.model.states.push({
                id: elemId,
                stateType: STATE_KEYWORD_MAP[keyword],
                identification: ident,
                label,
                placement,
                lineNumber: keywordToken.line,
                systemId: this.currentSystemId ?? undefined,
            } as State);
        } else if (keyword === 'process_operator') {
            this.model.processOperators.push({
                id: elemId,
                identification: ident,
                label,
                lineNumber: keywordToken.line,
                systemId: this.currentSystemId ?? undefined,
            } as ProcessOperator);
        } else if (keyword === 'technical_resource') {
            this.model.technicalResources.push({
                id: elemId,
                identification: ident,
                label,
                lineNumber: keywordToken.line,
                systemId: this.currentSystemId ?? undefined,
            } as TechnicalResource);
        }
    }

    // -- Connections --

    private parseConnection(): void {
        const sourceToken = this.current();
        const sourceId = sourceToken.value;
        this.advance();

        const token = this.current();
        if (token.type === TokenType.USAGE) {
            this.advance();
            if (!this.check(TokenType.IDENTIFIER)) {
                this.model.errors.push(
                    `Line ${sourceToken.line}: Expected identifier after connection operator`,
                );
                return;
            }
            const targetId = this.current().value;
            this.advance();

            if (!this.elementIds.has(sourceId)) {
                this.model.errors.push(
                    `Line ${sourceToken.line}: Element '${sourceId}' is not defined`,
                );
                return;
            }
            if (!this.elementIds.has(targetId)) {
                this.model.errors.push(
                    `Line ${sourceToken.line}: Element '${targetId}' is not defined`,
                );
                return;
            }

            this.usageCounter++;
            this.model.usages.push({
                id: `usage_${this.usageCounter}`,
                processOperatorRef: sourceId,
                technicalResourceRef: targetId,
                lineNumber: sourceToken.line,
                systemId: this.currentSystemId ?? undefined,
            } as Usage);
        } else if (FLOW_TOKEN_MAP.has(token.type)) {
            const flowType = FLOW_TOKEN_MAP.get(token.type)!;
            this.advance();

            if (!this.check(TokenType.IDENTIFIER)) {
                this.model.errors.push(
                    `Line ${sourceToken.line}: Expected identifier after connection operator`,
                );
                return;
            }
            const targetId = this.current().value;
            this.advance();

            if (!this.elementIds.has(sourceId)) {
                this.model.errors.push(
                    `Line ${sourceToken.line}: Element '${sourceId}' is not defined`,
                );
                return;
            }
            if (!this.elementIds.has(targetId)) {
                this.model.errors.push(
                    `Line ${sourceToken.line}: Element '${targetId}' is not defined`,
                );
                return;
            }

            this.flowCounter++;
            this.model.flows.push({
                id: `flow_${this.flowCounter}`,
                sourceRef: sourceId,
                targetRef: targetId,
                flowType,
                lineNumber: sourceToken.line,
                systemId: this.currentSystemId ?? undefined,
            } as Flow);
        } else {
            this.model.errors.push(
                `Line ${sourceToken.line}: Expected connection operator after '${sourceId}'`,
            );
        }
    }

    // -- Token helpers --

    private current(): Token {
        if (this.pos < this.tokens.length) {
            return this.tokens[this.pos];
        }
        return { type: TokenType.EOF, value: '', line: 0, column: 0 };
    }

    private check(tokenType: TokenType): boolean {
        return this.current().type === tokenType;
    }

    private advance(): Token {
        const token = this.current();
        if (this.pos < this.tokens.length) {
            this.pos++;
        }
        return token;
    }

    private expect(tokenType: TokenType): Token {
        if (this.check(tokenType)) {
            return this.advance();
        }
        const token = this.current();
        this.model.errors.push(`Line ${token.line}: Expected ${tokenType}, got ${token.type}`);
        return token;
    }

    private skipTrivial(): void {
        while (
            this.current().type === TokenType.NEWLINE ||
            this.current().type === TokenType.COMMENT
        ) {
            this.advance();
        }
    }
}
